processor.go

```diff
@@ -96,204 +116,193 @@ func (p *Processor[AC, OC, JC]) init() {
+	// One less job is executing for the prior state
+	p.stateStatusMap[state].Executing -= 1
+
+	// There are no waiting jobs for the state, so we have nothing to queue
+	waitingJobCount := len(p.stateWaitingJobsMap[state])
+	if waitingJobCount == 0 {
+		return
+	}
 
-		// flush out any new jobs we can
-		p.enqueueAllJobs(r)
+	job := p.stateWaitingJobsMap[state][waitingJobCount-1]
+	p.stateWaitingJobsMap[state] = p.stateWaitingJobsMap[state][0 : waitingJobCount-1]
+	p.stateStatusMap[job.State].Waiting -= 1
 
-		// If the state was terminal, we should see if all of the states are terminated, if so shut down
-		if !p.allJobsAreTerminal(r) {
-			continue
-		}
-		// if there are any jobs in flight in the run, keep going
-		if r.JobsInFlight() {
-			continue
-		}
+	p.runJob(job)
+}
 
+func (p *Processor[AC, OC, JC]) process(ctx context.Context, r *Run[OC, JC], wg *sync.WaitGroup) {
+	defer func() {
 		p.shutdown()
+		wg.Add(-1)
+	}()
 
-		break
+	// Enqueue the jobs to start
+	for _, job := range r.Jobs {
+		p.processJob(job)
 	}
-	slog.Info("ReturnChanWorker Quit")
-	wg.Done()
-}
 
-func (p *Processor[AC, OC, JC]) updateStatusCounts(r *Run[OC, JC]) {
-	counts := r.StatusCounts()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case completedJob := <-p.returnChan:
+			// If the prior state of the completed job was at capacity, we now have space for one more
+			p.runNextWaitingJob(completedJob.PriorState)
+
+			// Update the run with the new state
+			r.UpdateJob(completedJob.Job)
+			p.processJob(completedJob.Job)
+
+			// Start any of the new jobs that need kicking
+			for idx, kickRequest := range completedJob.KickRequests {
+				job := Job[JC]{
+					Id:          fmt.Sprintf("%s->%d", completedJob.Job.Id, idx),
+					C:           kickRequest.C,
+					State:       kickRequest.State,
+					StateErrors: map[string][]string{},
+				}
+				r.UpdateJob(job)
+				p.processJob(job)
+			}
 
-	ret := []StatusCount{}
+			if err := p.serializer.Serialize(*r); err != nil {
+				log.Fatalf("Error serializing, aborting now to not lose work: %v", err)
+			}
 
-	for _, state := range p.states {
-		if _, ok := counts[state.TriggerState]; !ok {
-			ret = append(ret, StatusCount{
-				State:     state.TriggerState,
-				Count:     0,
-				Executing: 0,
-				Terminal:  state.Terminal,
-			})
-			continue
+			p.updateStatus()
+
+			if p.allJobsAreTerminal(r) && !p.hasExecutingJobs() {
+				return
+			}
 		}
-		c := counts[state.TriggerState]
-		c.Terminal = state.Terminal
-		ret = append(ret, c)
 	}
-	p.statusListener.StatusUpdate(ret)
 }
 
-func (p *Processor[AC, OC, JC]) allJobsAreTerminal(r *Run[OC, JC]) bool {
-	c := r.StatusCounts()
-	for _, k := range p.states {
-		if k.Terminal {
-			continue
-		}
-		if c[k.TriggerState].Count > 0 {
-			return false
+func (p *Processor[AC, OC, JC]) canRunJobForState(state string) bool {
+	return p.stateStatusMap[state].Executing < p.stateMap[state].Concurrency
```

Comment:
mutex?

Reply:
the only code which can access this is non-concurrent (the single "manager" goroutine), so no synchronization primitives are necessary